File: Python3/0403. Frog Jump.py
class Solution:
    def canCross(self, stones: List[int]) -> bool:
        l = len(stones)
        dp = defaultdict(set)
        set_stone = set(stones)
        dp[0] = {0}
        for s in stones:
            for step in dp[s]:
                for d in [-1, 0, 1]:
                    if step + d > 0 and (s + step + d) in set_stone:
                        dp[s+ + step + d].add(step + d)
        return len(dp[stones[-1]]) > 0

File: C++/0403. Frog Jump.cpp
class Solution {
public:
    bool canCross(vector<int>& stones) {
        set<int> set_stones(stones.begin(), stones.end());
        map<int, set<int>> dp;
        dp[0].insert(0);

        for(auto s:set_stones) {
            for(auto step:dp[s]) {
                for(auto d:{1, 0, -1}) {
                    if((step + d > 0) && set_stones.count(s + step + d)) {
                        dp[s + step + d].insert(step + d);
                    }
                }
            }
        }

        return dp[stones.back()].size() != 0;
    }
};
